Specify codehash in plans

d-xo opened the pull request:
Since the introduction of the `CREATE2` opcode we can no longer assume that the same address will always contain the same code ([details](https://ethereum-magicians.org/t/potential-security-implications-of-create2-eip-1014/2614)).

Here, we defend against this attack by forcing callers of `plot` to specify the expected codehash of `usr` (the `tag`). When `exec` is called, the codehash is compared to the `tag`, and `exec` will revert if they do not match.

asymmetric commented on readme.md:
```diff
@@ -51,6 +52,7 @@ to security@dapp.org.
 **`exec`**
 - A `plan` can only be executed if it has previously been plotted
 - A `plan` can only be executed once it's `eta` has passed
+- A `plan` can only be executed if it's `tag` matches `extcodehash(usr)`
```
Its 😂

d-xo replied:
:woman_facepalming: :face_with_head_bandage:

asymmetric commented on src/pause.sol:
```diff
@@ -51,35 +51,43 @@ contract DSPause is DSAuth, DSNote {
 
-    // --- executions ---
-    function plot(address usr, bytes memory fax, uint eta)
+    function codehash(address usr)
```
Are you open to overloading `hash` for this?

Or do you prefer to keep it > 4?

d-xo replied:
I think it's a little weird since they would have quite different meanings semantically. I'm totally open to four letter naming suggestions.

d-xo replied:
I can think of:

- `code`
- `hash`
- `name`
- `sign`
- `mark`

But I don't love any of them. This function is annoying anyway tbh, since I only added it because for some reason solidity seems to only expose `extcodehash` in assembly.

asymmetric replied:
Do you know if they have any plans of surfacing it in Solidity? If so, it might not be worth it to rename it.

Anyway, here are a few more:

* `crux`
* `core`
* `gist`
* `pith`
* `soul`
* `quid`

These are all variations on the idea that the function gets you the essence of the code at `usr`.

d-xo replied:
`soul` is nice :), updated. Thanks! :pray:

asymmetric commented on src/pause.t.sol:
```diff
@@ -204,67 +218,86 @@ contract Exec is Test {
 
+    function testFail_tag_mismatch() public {
+        address      usr = target;
+        bytes32      tag = keccak256("hi");
```
It took me a while to spot this (and therefore to understand why the test is supposed to be failing) as someone less familiar with the code - maybe it would be clearer if you removed the call to `keccak256`?

d-xo replied:
updated.

d-xo committed:
Specify extcodehash in plans

Since the introduction of the CREATE2 opcode we can no longer assume
that the same address will always contain the same code
([details](https://ethereum-magicians.org/t/potential-security-implications-of-create2-eip-1014/2614)).

Here, we defend against this attack by forcing callers of plot to
specify the expected codehash of usr. When exec is called, the codehash
is compared to the tag, and exec will revert if they do not match.
